beacon-chain/light-client-network.md

```diff
@@ -0,0 +1,182 @@
+* The latest `LightClientOptimisticUpdate` known by a peer.
+
+## Specification
+
+### Distance Function
+
+The beacon chain light client network uses the stock XOR distance metric defined in the portal wire protocol specification.
+
+### Content ID Derivation Function
+
+The beacon chain light client network uses the SHA256 Content ID derivation function from the portal wire protocol specification.
+
+### Wire Protocol
+
+The [Portal wire protocol](./portal-wire-protocol.md) is used as the wire protocol for the Beacon Chain Light Client network.
+
+#### Protocol Identifier
+
+As specified in the [Protocol identifiers](./portal-wire-protocol.md#protocol-identifiers) section of the Portal wire protocol, the `protocol` field in the `TALKREQ` message **MUST** contain the value of `0x501A`.
+
+#### Supported Messages Types
+
+The beacon chain light client network supports the following protocol messages:
+
+- `Ping` - `Pong`
+- `Find Nodes` - `Nodes`
+- `Find Content` - `Found Content`
+- `Offer` - `Accept`
+
+#### `Ping.custom_data` & `Pong.custom_data`
+
+In the beacon chain light client network the `custom_payload` field of the `Ping` and `Pong` messages is the serialization of an SSZ Container specified as `custom_data`:
+
+```
+custom_data = Container(data_radius: uint256)
+custom_payload = serialize(custom_data)
+```
+
+### Routing Table
+
+The Beacon CHain Light Client Network uses the standard routing table structure from the Portal Wire Protocol.
+
+### Node State
+
+#### Data Radius
+
+The Beacon Chain Light Client Network includes one additional piece of node state that should be tracked. Nodes must track the `data_radius`
+from the Ping and Pong messages for other nodes in the network. This value is a 256 bit integer and represents the data that
+a node is "interested" in.
+
+We define the following function to determine whether node in the network should be interested in a piece of content:
+
+```
+interested(node, content) = distance(node.id, content.id) <= node.radius
+```
+
+A node is expected to maintain `radius` information for each node in its local node table. A node's `radius` value may fluctuate as the contents of its local key-value store change.
+
+A node should track their own radius value and provide this value in all Ping or Pong messages it sends to other nodes.
+
+### Data Types
+
+The beacon chain light client DHT stores the following data items:
+
+* LightClientBootstrap
+* LightClientUpdate
+
+The following data objects are ephemeral and we store only the latest values:
+
+* LightClientFinalityUpdate
+* LightClientOptimisticUpdate
+
+#### Constants
+
+We define the following constants which are used in the various data type definitions:
+
+```py
+# Maximum number of `LightClientUpdate` instances in a single request
+MAX_REQUEST_LIGHT_CLIENT_UPDATES = 2**7  # = 128
+```
+
+#### ForkDigest
+4-byte fork digest for the current beacon chain version and ``genesis_validators_root``.
+
+#### LightClientBootstrap
```

Comment:
At some point we will need to consider how we will inject bootstraps into the network.

Currently in Light Client protocol node choses one trusted block hash for which it will accept a bootstrap. If we would go that road, it would meant that when we would offer boostraps to the network most peers would only the one they trust, which would mean there is not a lot of bootstarp objects in the network.

As for now I see two ways to deal with it:
- make each node specify multiple trusted roots
- validate those out of band. (for example by libp2p light client ?)

Reply:
We also want to keep only bootstraps within the weak subjectivity period which is ~5 months, so we will need also a way to purge bootstraps for epochs < (current_epoch - MIN_EPOCHS_FOR_BLOCK_REQUESTS).

Reply:
IIUC this is a coordination issue, where we simply need the clients to "agree" on which blocks we will consider valid for "bootstrapping".  Assuming the selection criteria is arbitrary, can we just define something like a rolling window/epoch of ~2 months worth of blocks.  This would ensure that we always had at least 2 block hashes that could be used for bootstrapping within the previous 5 month time window?